### DISCUSSION TITLE
DeepSpeed requires the LOCAL_RANK environment variable

### DISCUSSION THREAD

**Discussion by nickKyr** [deepspeedai/DeepSpeed#1242] [Q&A] [answered]
Hello,

I am trying to use DeepSpeed in the `pl.Trainer` of PyTorch Lightning and I get the following error:
_AssertionError: DeepSpeed requires the LOCAL_RANK environment variable, it is set by the deepspeed launcher, deepspeed.init_distributed, or the torch.distributed launcher. If using a different launcher please ensure LOCAL_RANK is set prior to initializing deepspeed._

The trainer is:        
 ```
trainer = pl.Trainer(
            gpus=2, accelerator='ddp',
            callbacks=[checkpoint, callback1],
            max_epochs=epochs,
            logger=logger1,
            plugins=[DeepSpeedPlugin()], precision=16
            )
```
and the corresponding import at the beginning of the file:
`from pytorch_lightning.plugins import DeepSpeedPlugin`
Is the above problem related to [Distributed Initialization](https://deepspeed.readthedocs.io/en/stable/initialize.html?highlight=torch%20distributed#distributed-initialization) and therefore, `deepspeed.init_distributed(init_method="Torch distributed")` must be added prior to the `Trainer`?

Any other suggestion is more than welcome.

Thanks,
Nikos

**Comment by tjruwase** [comment]
@nickKyr, is this a general problem with DeepSpeed+Lightning or specific to your scenario? 
@SeanNaren for awareness.

**Reply by nickKyr** [reply]
Hi,
not sure, I guess it is something related to my application, as it is the first time I am using DeepSpeed. That was basically the main purpose of my question. To check if you have seen this error before and if you have noticed any problems with DeepSpeed+Pytorch Lightning.

It seems like the `LOCAL_RANK` must be explicitly set. I followed the [example](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/tests/plugins/test_deepspeed_plugin.py) and I added the parts shown in the post above, but got this error :/

Thanks,
Nikos

**Reply by tjruwase** [reply]
@nickKyr, got it. Yes, `LOCAL_RANK` needs to be set for the process to know which GPU to use. And this is normally done by the various distributed launchers. We have not previously seen this problem with DeepSpeed+Pytorch Lightning. How are you launching your run?

**Reply by nickKyr** [reply]
I submit my run with slurm queueing system with the following script:
```
# Slurm job options (name, compute nodes, job time)
#SBATCH --job-name=AIpred
#SBATCH --time=2:00:00
#SBATCH --partition=gpu-cascade
#SBATCH --qos=gpu
#SBATCH --gres=gpu:2

# Load the default HPC MPI environment
module load anaconda/python3
module load nvidia/cuda-10.2
source ~/minicondaInit.sh

export OMP_NUM_THREADS=1

srun python3 train.py 2>&1 | tee log.train
```
Based on [Distributed Initialization](https://deepspeed.readthedocs.io/en/stable/initialize.html?highlight=torch%20distributed#distributed-initialization), the default value is “env://” and I don't explicitly define LOCAL_RANK in the above job submission file. Should I define the system LOCAL_RANK or the Pytorch distributed: `deepspeed.init_distributed(init_method="Torch distributed")`? Is there an example available how to do that?

In addition, if I try to run it on my local machine (1 GPU), I get the following error:
_RuntimeError: Tensors must be CUDA and dense_ in line: `self.log('epoch', self.trainer.current_epoch, sync_dist=True)`.
I get the above error even for a similar plugin in the `pl.Trainer`: `plugins=DeepSpeedPlugin(stage=3, cpu_offload=True), precision=16`. I assume the tensor is transformed from dense to sparse when using DeepSpeed?
Have you seen this error before ([same error output](https://github.com/PyTorchLightning/pytorch-lightning/issues/2205) doesn't seem relevant)?

Thanks,
Nikos

**Reply by tjruwase** [reply]
@nickKyr, I have not seen this RuntimeError previously. I am not sure what you mean by tensor transformed from dense to spare. What is not clear to me now is whether this is an issue with DeepSpeed, Lightning, or the integration. I am also not very familiar with the slurm environment. A couple of things to try:

1. Can you run with Lightning alone, without DeepSpeed? Even if you have to reduce model/batch size.
2. Can you run with DeepSpeed stage 1 and 2?

**Reply by nickKyr** [reply]
Hi @tjruwase , thanks for the reply.
1. Yes, the code works with PyTorch Lightning both on my local machine and on the GPU cluster (with slurm queueing system). I wanted to integrate DeepSpeed in my PyTorch Lightning workflow in order to train higher resolution samples. So, the modification I made in the working code was to add the arguments `plugins=DeepSpeedPlugin(), precision=16` inside `pl.Trainer` call and I imported the corresponding plugin: ` from pytorch_lightning.plugins import DeepSpeedPlugin`, as suggested in [the documentation](https://pytorch-lightning.readthedocs.io/en/latest/advanced/advanced_gpu.html#deepspeed). When I try to train the model, I get the errors posted above...
2. Yes, I have tried stages 1, 2 and 3, all give the same error:
_AssertionError: DeepSpeed requires the LOCAL_RANK environment variable, it is set by the deepspeed launcher, deepspeed.init_distributed, or the torch.distributed launcher. If using a different launcher please ensure LOCAL_RANK is set prior to initializing deepspeed._
The error comes from line 599 in deepspeed/runtime/engine.py.

I tried to look into `DeepSpeedPlugin` class, which has been created in _site-packages/pytorch_lightning/plugins/training_type_, and it seems that it inherits from `DDPPlugin`, so I don't understand why the devices and/or the cluster environment are not recognized, while ddp was working.

**Reply by SeanNaren** [reply]
hey @nickKyr thanks for the issue and thank @tjruwase for helping!

What version of Lightning are you using? Can you by any chance use 1.4? (`pip install pytorch-lightning -U`). Recently I modified the code to completely rely on the deepspeed init_distributed, setting variables as needed. Might work now!

**Reply by nickKyr** [reply]
hi @SeanNaren and thanks for the reply. Indeed, using Lightning v1.4 solved the problem initially, DeepSpeed is initialized and the GPUs are recognized:

_initializing deepspeed distributed: GLOBAL_RANK: 0, MEMBER: 1/2
initializing deepspeed distributed: GLOBAL_RANK: 1, MEMBER: 2/2
Enabling DeepSpeed FP16.
LOCAL_RANK: 0 - CUDA_VISIBLE_DEVICES: [0,1]
LOCAL_RANK: 1 - CUDA_VISIBLE_DEVICES: [0,1]_

However, before starting the first epoch, the following error appears:
_RuntimeError: Input type (torch.cuda.FloatTensor) and weight type (torch.cuda.HalfTensor) should be the same._
It is related to the line of the `forward` method where the input Tensor (of type float) is inserted into the model. Therefore, I modified accordingly the `training_step` and the `validation_step` methods:
```
    def training_step(self, batch, batch_idx):
        inputs, targets = batch
        inputs, targets = inputs.half(), targets.half() #inputs, targets = inputs.float(), targets.float()
        outputs = self.forward(inputs)
```
Then I get the error below which is confusing:

  File "/miniconda3/lib/python3.7/site-packages/torch/nn/modules/conv.py", line 587, in forward return self._conv_forward(input, self.weight, self.bias)_
  File "/miniconda3/lib/python3.7/site-packages/torch/nn/modules/conv.py", line 583, in _conv_forward input, weight, bias, self.stride, self.padding, self.dilation, self.groups_
_TypeError: conv3d() received an invalid combination of arguments - got (builtin_function_or_method, Parameter, Parameter, tuple, tuple, tuple, int), but expected one of:_
 * _(Tensor input, Tensor weight, Tensor bias, tuple of ints stride, tuple of ints padding, tuple of ints dilation, int groups)
      didn't match because some of the arguments have invalid types: (!builtin_function_or_method!, !Parameter!, !Parameter!, !tuple!, !tuple!, !tuple!, int)_
 * _(Tensor input, Tensor weight, Tensor bias, tuple of ints stride, str padding, tuple of ints dilation, int groups)
      didn't match because some of the arguments have invalid types: (!builtin_function_or_method!, !Parameter!, !Parameter!, !tuple!, !tuple!, !tuple!, int)_

I don't think there is something wrong with the convolution operations as they were working without the DeepSpeed plugin. There must be a mismatch between mixed precision and the inputs/outputs which are floats(32 bit). Please note that the above have been noticed when running on 2 GPUs on a cluster system. If I try to run it on my local machine, the errors are the same, apart from the last one. Instead of the error in conv3d, I get the following:

  _File "/netType.py", line 317, in validation_epoch_end
    self.log('epoch', self.trainer.current_epoch, sync_dist=True)
  File "/anaconda3/lib/python3.8/site-packages/pytorch_lightning/core/lightning.py", line 286, in log
    self._results.log(
  File "/anaconda3/lib/python3.8/site-packages/pytorch_lightning/core/step_result.py", line 149, in log
    value = sync_fn(value, group=sync_dist_group, reduce_op=sync_dist_op)
  File "/anaconda3/lib/python3.8/site-packages/pytorch_lightning/plugins/training_type/ddp.py", line 290, in reduce
    output = sync_ddp_if_available(output, group, reduce_op)
  File "/anaconda3/lib/python3.8/site-packages/pytorch_lightning/utilities/distributed.py", line 129, in sync_ddp_if_available
    return sync_ddp(result, group=group, reduce_op=reduce_op)
  File "/anaconda3/lib/python3.8/site-packages/pytorch_lightning/utilities/distributed.py", line 162, in sync_ddp
    torch.distributed.all_reduce(result, op=op, group=group, async_op=False)
  File "/anaconda3/lib/python3.8/site-packages/torch/distributed/distributed_c10d.py", line 1171, in all_reduce
    work = group.allreduce([tensor], opts)
RuntimeError: Tensors must be CUDA and dense_/

Any ideas?
Thanks,
Nikos

**Reply by tjruwase** [reply]
@nickKyr, yes this could a mixed-precision issue since those can be tricky to get right. Can you try running purely fp32 mode? To my knowledge DeepSpeed plugin can support that, you can see [this](https://github.com/microsoft/DeepSpeed/issues/1233) for more details.

**Reply by nickKyr** [reply]
@tjruwase that worked! Thank you both @tjruwase and @SeanNaren for the combined effort and your ideas! 
I had read somewhere that DeepSpeed can run in 32 mode, but I was getting an error that precision must be set to 16 (with older Lightning version). It wasn't until the most recent PyTorch Lightning v1.4 that this was fixed based on the link you sent and I hadn't tried this with v1.4...

To sum up, Lightning v1.4 is required to integrate DeepSpeed. The only changes needed are the addition of the corresponding arguments in the trainer call. On a similar note, I noticed that the order of `setup` and `train_dataloader` methods is reversed when integrating DeepSpeed in Lightning.

Cheers,
Nikos

**Reply by tjruwase** [reply]
@nickKyr, thanks for sharing this update. I am so glad to hear that Lightning+DeepSpeed is now working for you. That is awesome. 

It does seem like some work is needed to make mixed-precision training a smoother process. But perhaps, mixed-precision is not an urgent need for you at this time. But whenever you do plan to try mixed-precision, do let us know and we can solve the issues together :).

**Reply by nickKyr** [reply]
Hi, mixed-precision seems to work in Lightning v1.4 as well! I just added `precision=16` in the `trainer` and modified the inputs and targets, which are actually float, in `training_step` and `validation_step`:
```
    def training_step(self, batch, batch_idx):
        inputs, targets = batch
        inputs, targets = inputs.half(), targets.half()
        outputs = self.forward(inputs)
        ...
```

**Reply by tjruwase** [reply]
Thanks for sharing that update. That is absolutely awesome to hear. Big kudos to @SeanNaren and co for enabling a seamless mixed-precision training experience in Lightning. That is no trivial achievement. 

@nickKyr, do you think we can now close this question as answered?

**Reply by nickKyr** [reply]
Yes, the issue has been resolved. Thanks for the new capabilities that you have added!

**Comment by tjruwase** [comment, marked as answer]
`LOCAL_RANK` (and mixed-precision) issue resolved by upgrading to Lightning v1.4.

**Comment by shoang22** [comment]
Hello,

I'm running into a similar issue. I'm trying to run DeepSpeed (0.3.12) + Lightning (1.2.3) in the following toy example `train_boring.py` using a SLURM cluster:

```
import os
import torch
from torch.utils.data import DataLoader, Dataset

from pytorch_lightning import LightningModule, Trainer
from pytorch_lightning.plugins import DeepSpeedPlugin


class RandomDataset(Dataset):
    def __init__(self, size, length):
        self.len = length
        self.data = torch.randn(length, size)

    def __getitem__(self, index):
        return self.data[index]

    def __len__(self):
        return self.len


class BoringModel(LightningModule):
    def __init__(self):
        super().__init__()
        self.layer = torch.nn.Linear(32, 2)

    def forward(self, x):
        return self.layer(x)

    def training_step(self, batch, batch_idx):
        loss = self(batch).sum()
        self.log("train_loss", loss)
        return {"loss": loss}

    def validation_step(self, batch, batch_idx):
        loss = self(batch).sum()
        self.log("valid_loss", loss)

    def test_step(self, batch, batch_idx):
        loss = self(batch).sum()
        self.log("test_loss", loss)

    def configure_optimizers(self):
        return torch.optim.SGD(self.layer.parameters(), lr=0.1)


def run():
    train_data = DataLoader(RandomDataset(32, 64), batch_size=2)
    val_data = DataLoader(RandomDataset(32, 64), batch_size=2)
    test_data = DataLoader(RandomDataset(32, 64), batch_size=2)

    ngpus = int(os.environ["SLURM_NTASKS_PER_NODE"])
    nnodes = int(os.environ["SLURM_NNODES"])

    ds_config = {
        "zero_optimization": {
            "stage": 2,
            "cpu_offload": False,
            "contiguous_gradients": True,
            "overlap_comm": True,
            "allgather_bucket_size": 5e8,
            "reduce_bucket_size": 5e8
        },
        "fp16": {
            "enabled": True,
            "loss_scale": 0.0,
            "initial_scale_power": 32,
            "loss_scale_window": 1000,
            "hysteresis": 2,
            "min_loss_scale": 1000
        }
    }

    plugins = [DeepSpeedPlugin(ds_config)]

    print("N GPUS: {}".format(ngpus))
    print("N NODES: {}".format(nnodes))

    model = BoringModel()
    trainer = Trainer(
        accelerator="ddp",
        gpus=ngpus,
        num_nodes=nnodes,
        default_root_dir=os.getcwd(),
        limit_train_batches=1,
        limit_val_batches=1,
        limit_test_batches=1,
        num_sanity_val_steps=0,
        max_epochs=1,
        plugins=plugins,
        precision=16
    )
    trainer.fit(model, train_data)
    trainer.test(model, test_data)


if __name__ == "__main__":
    run()
```

The following script is used to launch the SLURM job:
```
#!/bin/bash
#SBATCH --job-name=boring
#SBATCH --nodes=1
#SBATCH --ntasks-per-node=4
#SBATCH -p gpu
#SBATCH --gres=gpu:4
#SBATCH --mem=64gb
#SBATCH --time=0:10:00
#SBATCH --output=slurm_out/output.%A_%a.txt

module load miniconda3
module load cuda/11.8.0
nvidia-smi

export MASTER_PORT=$(expr 10000 + $(echo -n $SLURM_JOBID | tail -c 4))
master_addr=$(scontrol show hostnames "$SLURM_JOB_NODELIST" | head -n 1)
export MASTER_ADDR=$master_addr
echo "MASTER_ADDR="$MASTER_ADDR    

srun python train_boring.py
```

executing this results in the error below:

```
+---------------------------------------------------------------------------------------+
| NVIDIA-SMI 535.104.05             Driver Version: 535.104.05   CUDA Version: 12.2     |
|-----------------------------------------+----------------------+----------------------+
| GPU  Name                 Persistence-M | Bus-Id        Disp.A | Volatile Uncorr. ECC |
| Fan  Temp   Perf          Pwr:Usage/Cap |         Memory-Usage | GPU-Util  Compute M. |
|                                         |                      |               MIG M. |
|=========================================+======================+======================|
|   0  Tesla V100-SXM2-32GB           On  | 00000000:1A:00.0 Off |                    0 |
| N/A   30C    P0              43W / 300W |      0MiB / 32768MiB |      0%   E. Process |
|                                         |                      |                  N/A |
+-----------------------------------------+----------------------+----------------------+
|   1  Tesla V100-SXM2-32GB           On  | 00000000:1C:00.0 Off |                    0 |
| N/A   30C    P0              43W / 300W |      0MiB / 32768MiB |      0%   E. Process |
|                                         |                      |                  N/A |
+-----------------------------------------+----------------------+----------------------+
|   2  Tesla V100-SXM2-32GB           On  | 00000000:1D:00.0 Off |                    0 |
| N/A   29C    P0              41W / 300W |      0MiB / 32768MiB |      0%   E. Process |
|                                         |                      |                  N/A |
+-----------------------------------------+----------------------+----------------------+
|   3  Tesla V100-SXM2-32GB           On  | 00000000:1E:00.0 Off |                    0 |
| N/A   32C    P0              43W / 300W |      0MiB / 32768MiB |      0%   E. Process |
|                                         |                      |                  N/A |
+-----------------------------------------+----------------------+----------------------+
                                                                                         
+---------------------------------------------------------------------------------------+
| Processes:                                                                            |
|  GPU   GI   CI        PID   Type   Process name                            GPU Memory |
|        ID   ID                                                             Usage      |
|=======================================================================================|
|  No running processes found                                                           |
+---------------------------------------------------------------------------------------+
MASTER_ADDR=mg090
GpuFreq=control_disabled
INFO:    underlay of /etc/localtime required more than 50 (80) bind mounts
INFO:    underlay of /usr/bin/nvidia-smi required more than 50 (251) bind mounts
INFO:    underlay of /etc/localtime required more than 50 (80) bind mounts
INFO:    underlay of /etc/localtime required more than 50 (80) bind mounts
INFO:    underlay of /etc/localtime required more than 50 (80) bind mounts
INFO:    underlay of /usr/bin/nvidia-smi required more than 50 (251) bind mounts
INFO:    underlay of /usr/bin/nvidia-smi required more than 50 (251) bind mounts
INFO:    underlay of /usr/bin/nvidia-smi required more than 50 (251) bind mounts
GPU available: True, used: True
TPU available: None, using: 0 TPU cores
Multi-processing is handled by Slurm.
GPU available: True, used: True
TPU available: None, using: 0 TPU cores
Multi-processing is handled by Slurm.
GPU available: True, used: True
TPU available: None, using: 0 TPU cores
Multi-processing is handled by Slurm.
GPU available: True, used: True
TPU available: None, using: 0 TPU cores
Multi-processing is handled by Slurm.
/opt/conda/lib/python3.8/site-packages/pytorch_lightning/utilities/distributed.py:50: UserWarning: you defined a validation_step but have no val_dataloader. Skipping validation loop
  warnings.warn(*args, **kwargs)
initializing ddp: GLOBAL_RANK: 3, MEMBER: 4/4
/opt/conda/lib/python3.8/site-packages/pytorch_lightning/utilities/distributed.py:50: UserWarning: you defined a validation_step but have no val_dataloader. Skipping validation loop
  warnings.warn(*args, **kwargs)
initializing ddp: GLOBAL_RANK: 1, MEMBER: 2/4
/opt/conda/lib/python3.8/site-packages/pytorch_lightning/utilities/distributed.py:50: UserWarning: you defined a validation_step but have no val_dataloader. Skipping validation loop
  warnings.warn(*args, **kwargs)
initializing ddp: GLOBAL_RANK: 2, MEMBER: 3/4
/opt/conda/lib/python3.8/site-packages/pytorch_lightning/utilities/distributed.py:50: UserWarning: you defined a validation_step but have no val_dataloader. Skipping validation loop
  warnings.warn(*args, **kwargs)
initializing ddp: GLOBAL_RANK: 0, MEMBER: 1/4
You have not specified an optimizer or scheduler within the DeepSpeed config.Using `configure_optimizers` to define optimizer and scheduler.
INFO:lightning:You have not specified an optimizer or scheduler within the DeepSpeed config.Using `configure_optimizers` to define optimizer and scheduler.
Traceback (most recent call last):
  File "/opt/conda/lib/python3.8/runpy.py", line 194, in _run_module_as_main
    return _run_code(code, main_globals, None,
  File "/opt/conda/lib/python3.8/runpy.py", line 87, in _run_code
    exec(code, run_globals)
  File "/ui/abv/hoangsx/Chemformer/temp/train_boring.py", line 100, in <module>
    run()
  File "/ui/abv/hoangsx/Chemformer/temp/train_boring.py", line 95, in run
    trainer.fit(model, train_data)
  File "/opt/conda/lib/python3.8/site-packages/pytorch_lightning/trainer/trainer.py", line 511, in fit
    self.pre_dispatch()
  File "/opt/conda/lib/python3.8/site-packages/pytorch_lightning/trainer/trainer.py", line 540, in pre_dispatch
    self.accelerator.pre_dispatch()
  File "/opt/conda/lib/python3.8/site-packages/pytorch_lightning/accelerators/accelerator.py", line 84, in pre_dispatch
    self.training_type_plugin.pre_dispatch()
  File "/opt/conda/lib/python3.8/site-packages/pytorch_lightning/plugins/training_type/deepspeed.py", line 174, in pre_dispatch
    self.init_deepspeed()
  File "/opt/conda/lib/python3.8/site-packages/pytorch_lightning/plugins/training_type/deepspeed.py", line 193, in init_deepspeed
    self._initialize_deepspeed_train(model)
  File "/opt/conda/lib/python3.8/site-packages/pytorch_lightning/plugins/training_type/deepspeed.py", line 218, in _initialize_deepspeed_train
    model, optimizer, _, lr_scheduler = deepspeed.initialize(
  File "/opt/conda/lib/python3.8/site-packages/deepspeed/__init__.py", line 112, in initialize
    engine = DeepSpeedEngine(args=args,
  File "/opt/conda/lib/python3.8/site-packages/deepspeed/runtime/engine.py", line 145, in __init__
    self._configure_with_arguments(args, mpu)
  File "/opt/conda/lib/python3.8/site-packages/deepspeed/runtime/engine.py", line 499, in _configure_with_arguments
    args.local_rank = int(os.environ['LOCAL_RANK'])
  File "/opt/conda/lib/python3.8/os.py", line 675, in __getitem__
    raise KeyError(key) from None
KeyError: 'LOCAL_RANK'
N GPUS: 4
N NODES: 1
You have not specified an optimizer or scheduler within the DeepSpeed config.Using `configure_optimizers` to define optimizer and scheduler.
INFO:lightning:You have not specified an optimizer or scheduler within the DeepSpeed config.Using `configure_optimizers` to define optimizer and scheduler.
You have not specified an optimizer or scheduler within the DeepSpeed config.Using `configure_optimizers` to define optimizer and scheduler.
INFO:lightning:You have not specified an optimizer or scheduler within the DeepSpeed config.Using `configure_optimizers` to define optimizer and scheduler.
You have not specified an optimizer or scheduler within the DeepSpeed config.Using `configure_optimizers` to define optimizer and scheduler.
INFO:lightning:You have not specified an optimizer or scheduler within the DeepSpeed config.Using `configure_optimizers` to define optimizer and scheduler.
Traceback (most recent call last):
  File "/opt/conda/lib/python3.8/runpy.py", line 194, in _run_module_as_main
    return _run_code(code, main_globals, None,
  File "/opt/conda/lib/python3.8/runpy.py", line 87, in _run_code
    exec(code, run_globals)
  File "/ui/abv/hoangsx/Chemformer/temp/train_boring.py", line 100, in <module>
    run()
  File "/ui/abv/hoangsx/Chemformer/temp/train_boring.py", line 95, in run
    trainer.fit(model, train_data)
  File "/opt/conda/lib/python3.8/site-packages/pytorch_lightning/trainer/trainer.py", line 511, in fit
    self.pre_dispatch()
  File "/opt/conda/lib/python3.8/site-packages/pytorch_lightning/trainer/trainer.py", line 540, in pre_dispatch
    self.accelerator.pre_dispatch()
  File "/opt/conda/lib/python3.8/site-packages/pytorch_lightning/accelerators/accelerator.py", line 84, in pre_dispatch
    self.training_type_plugin.pre_dispatch()
  File "/opt/conda/lib/python3.8/site-packages/pytorch_lightning/plugins/training_type/deepspeed.py", line 174, in pre_dispatch
    self.init_deepspeed()
  File "/opt/conda/lib/python3.8/site-packages/pytorch_lightning/plugins/training_type/deepspeed.py", line 193, in init_deepspeed
    self._initialize_deepspeed_train(model)
  File "/opt/conda/lib/python3.8/site-packages/pytorch_lightning/plugins/training_type/deepspeed.py", line 218, in _initialize_deepspeed_train
    model, optimizer, _, lr_scheduler = deepspeed.initialize(
  File "/opt/conda/lib/python3.8/site-packages/deepspeed/__init__.py", line 112, in initialize
    engine = DeepSpeedEngine(args=args,
  File "/opt/conda/lib/python3.8/site-packages/deepspeed/runtime/engine.py", line 145, in __init__
    self._configure_with_arguments(args, mpu)
  File "/opt/conda/lib/python3.8/site-packages/deepspeed/runtime/engine.py", line 499, in _configure_with_arguments
    args.local_rank = int(os.environ['LOCAL_RANK'])
  File "/opt/conda/lib/python3.8/os.py", line 675, in __getitem__
    raise KeyError(key) from None
KeyError: 'LOCAL_RANK'
N GPUS: 4
N NODES: 1
Traceback (most recent call last):
  File "/opt/conda/lib/python3.8/runpy.py", line 194, in _run_module_as_main
    return _run_code(code, main_globals, None,
  File "/opt/conda/lib/python3.8/runpy.py", line 87, in _run_code
    exec(code, run_globals)
  File "/ui/abv/hoangsx/Chemformer/temp/train_boring.py", line 100, in <module>
    run()
  File "/ui/abv/hoangsx/Chemformer/temp/train_boring.py", line 95, in run
    trainer.fit(model, train_data)
  File "/opt/conda/lib/python3.8/site-packages/pytorch_lightning/trainer/trainer.py", line 511, in fit
    self.pre_dispatch()
  File "/opt/conda/lib/python3.8/site-packages/pytorch_lightning/trainer/trainer.py", line 540, in pre_dispatch
    self.accelerator.pre_dispatch()
  File "/opt/conda/lib/python3.8/site-packages/pytorch_lightning/accelerators/accelerator.py", line 84, in pre_dispatch
    self.training_type_plugin.pre_dispatch()
  File "/opt/conda/lib/python3.8/site-packages/pytorch_lightning/plugins/training_type/deepspeed.py", line 174, in pre_dispatch
    self.init_deepspeed()
  File "/opt/conda/lib/python3.8/site-packages/pytorch_lightning/plugins/training_type/deepspeed.py", line 193, in init_deepspeed
    self._initialize_deepspeed_train(model)
  File "/opt/conda/lib/python3.8/site-packages/pytorch_lightning/plugins/training_type/deepspeed.py", line 218, in _initialize_deepspeed_train
    model, optimizer, _, lr_scheduler = deepspeed.initialize(
  File "/opt/conda/lib/python3.8/site-packages/deepspeed/__init__.py", line 112, in initialize
    engine = DeepSpeedEngine(args=args,
  File "/opt/conda/lib/python3.8/site-packages/deepspeed/runtime/engine.py", line 145, in __init__
    self._configure_with_arguments(args, mpu)
  File "/opt/conda/lib/python3.8/site-packages/deepspeed/runtime/engine.py", line 499, in _configure_with_arguments
    args.local_rank = int(os.environ['LOCAL_RANK'])
  File "/opt/conda/lib/python3.8/os.py", line 675, in __getitem__
    raise KeyError(key) from None
KeyError: 'LOCAL_RANK'
N GPUS: 4
N NODES: 1
N GPUS: 4
N NODES: 1
Traceback (most recent call last):
  File "/opt/conda/lib/python3.8/runpy.py", line 194, in _run_module_as_main
    return _run_code(code, main_globals, None,
  File "/opt/conda/lib/python3.8/runpy.py", line 87, in _run_code
    exec(code, run_globals)
  File "/ui/abv/hoangsx/Chemformer/temp/train_boring.py", line 100, in <module>
    run()
  File "/ui/abv/hoangsx/Chemformer/temp/train_boring.py", line 95, in run
    trainer.fit(model, train_data)
  File "/opt/conda/lib/python3.8/site-packages/pytorch_lightning/trainer/trainer.py", line 511, in fit
    self.pre_dispatch()
  File "/opt/conda/lib/python3.8/site-packages/pytorch_lightning/trainer/trainer.py", line 540, in pre_dispatch
    self.accelerator.pre_dispatch()
  File "/opt/conda/lib/python3.8/site-packages/pytorch_lightning/accelerators/accelerator.py", line 84, in pre_dispatch
    self.training_type_plugin.pre_dispatch()
  File "/opt/conda/lib/python3.8/site-packages/pytorch_lightning/plugins/training_type/deepspeed.py", line 174, in pre_dispatch
    self.init_deepspeed()
  File "/opt/conda/lib/python3.8/site-packages/pytorch_lightning/plugins/training_type/deepspeed.py", line 193, in init_deepspeed
    self._initialize_deepspeed_train(model)
  File "/opt/conda/lib/python3.8/site-packages/pytorch_lightning/plugins/training_type/deepspeed.py", line 218, in _initialize_deepspeed_train
    model, optimizer, _, lr_scheduler = deepspeed.initialize(
  File "/opt/conda/lib/python3.8/site-packages/deepspeed/__init__.py", line 112, in initialize
    engine = DeepSpeedEngine(args=args,
  File "/opt/conda/lib/python3.8/site-packages/deepspeed/runtime/engine.py", line 145, in __init__
    self._configure_with_arguments(args, mpu)
  File "/opt/conda/lib/python3.8/site-packages/deepspeed/runtime/engine.py", line 499, in _configure_with_arguments
    args.local_rank = int(os.environ['LOCAL_RANK'])
  File "/opt/conda/lib/python3.8/os.py", line 675, in __getitem__
    raise KeyError(key) from None
KeyError: 'LOCAL_RANK'
```

Attempting to upgrade to Pytorch Lightning 1.4.0 resulted in the same error. Is there something I'm missing?
